itensor/decomp.cc

```diff
@@ -925,4 +925,185 @@ expHermitian(ITensor const& T, Cplx t)
+    for(const auto& I : AA.inds())
+        {
+        if(hasIndex(A,I)) Ainds.push_back(I);
+        else              Binds.push_back(I);
+        }
+    ITensor Qcomb,
+            Rcomb;
+    Index qi,
+          ri;
+
+    auto AAcomb = AA;
+    if(!Qinds.empty())
+        {
+        std::tie(Qcomb,qi) = combiner(std::move(Qinds));
+        AAcomb *= Qcomb;
+        }
+    if(!Rinds.empty())
+        {
+        std::tie(Rcomb,ri) = combiner(std::move(Rinds));
+        AAcomb *= Rcomb;
+        }
+    
+    qr_decompOrd2(AAcomb,qi,ri,Q,R,args);
+
+    Q = dag(Qcomb) * Q;
+    R = R * dag(Rcomb);
+    } //qr_decomp
+
+std::tuple<ITensor,ITensor>
+qr_decomp(ITensor const& AA, IndexSet const& Qis, IndexSet const& Ris,
+    Args args)
+    {
+    if( !hasSameInds(inds(AA),IndexSet(Qis,Ris)) )
+      Error("In QR, Q indices and R indices must match the indices of the input ITensor");
+    return qr_decomp(AA,Qis,args);
+    }
+
+std::tuple<ITensor,ITensor>
+qr_decomp(ITensor const& AA, IndexSet const& Qis,
+    Args args)
+    {
+    ITensor Q(Qis),R;
+    qr_decomp(AA,Q,R,args);
+    auto q = commonIndex(Q,R);
+    return std::tuple<ITensor,ITensor>(Q,R);
+    }
+
+template<typename T>
+void
+qrImpl(ITensor const& A,
+        Index const& qI, 
+        Index const& rI,
+        ITensor & Q,
+        ITensor & R,
+        Args args)
+    {
+   
+    auto internaltagset = getTagSet(args,"InternalTags","Link,QR");
+    
+    if(not hasQNs(A))
+        {
+        auto matA = toMatRefc<T>(A,qI,rI);
+
+        Mat<T> QQ,RR;
+
+        QR(matA, QQ, RR);
+        
+        auto qL = Index(nrows(matA),internaltagset);
+        auto rL = setTags(qL,internaltagset);
+
+        Q = ITensor({qI,qL},Dense<T>(move(QQ.storage())));
+        R = ITensor({rL,rI},Dense<T>(move(RR.storage())));
+	}
+    else
+        {
+        auto blocks = doTask(GetBlocks<T>{A.inds(),qI,rI},A.store());
+        auto Nblock = blocks.size();
+        if(Nblock == 0) throw ResultIsZero("IQTensor has no blocks");
+	auto Liq = Index::qnstorage{};
+        auto Riq = Index::qnstorage{};
+        Liq.reserve(Nblock);
+        Riq.reserve(Nblock);
+
+
+        //TODO: optimize allocation/lookup of Qmats,Rmats
+        //      etc. by allocating memory ahead of time (see algs.cc)
+        //      and making Qmats a vector of MatrixRef's to this memory
+        auto Qmats = vector<Mat<T>>(Nblock);
+        auto Rmats = vector<Mat<T>>(Nblock);
+	for(auto b : range(Nblock))
+            {
+	      auto& B = blocks[b];
+	      auto& matA = B.M;
+
+	      auto & QQ = Qmats.at(b);
+	      auto & RR = Rmats.at(b);
+	      QR(matA, QQ, RR);
+	      Liq.emplace_back(qn(qI,1+B.i1),nrows(matA));
+	    }
+	
+	auto qL = Index(move(Liq),qI.dir(),internaltagset);
+	auto rL = qL;
+	rL.setDir(rI.dir());
+	rL.setTags(internaltagset);
+	auto Qis = IndexSet(qI,dag(qL));
+        auto Ris = IndexSet(rL, rI);
+
+        auto Qstore = QDense<T>(Qis,QN());
+        auto Rstore = QDense<T>(Ris,QN());
```

Comment:
There will be some cases where the flux (total QN) of the original tensor is not equal to zero, in which case R should absorb this flux. (Q should always be flux zero since it is unitary i.e. just a change of basis.) 

It would be great if you could add a unit test which tests a non-zero flux tensor T and checks that its QR decomp works (I think it won't unless you propagate T's flux here to R).